virtcontainers/ich.go

```diff
@@ -0,0 +1,813 @@
+	ichSocket             = "ich.sock"
+	vfsdSocket            = "virtiofsd.sock"
+	apiSocket             = "cloud-hypervisor.9166"
+	ichStopSandboxTimeout = 15
+)
+
+// the info struct goes to disk as hypervisor.json!
+
+type CloudHypervisorInfo struct {
+	PID int
+}
+
+//
+// cloud hypervisor state
+//
+type CloudHypervisorState struct {
+	sync.RWMutex
+	state ichState
+	PID int
+	VirtiofsdPid int
+	UUID string
+}
+
+func (s *CloudHypervisorState) set(state ichState) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.state = state
+}
+
+//
+// cloud hypervisor 
+//
+type cloudHypervisor struct {
+	id         string 
+	state      CloudHypervisorState
+	info       CloudHypervisorInfo
+	ichd       *exec.Cmd
+	store      *store.VCStore
+	config     HypervisorConfig
+	ctx        context.Context
+	socketPath string
+	ichConfig  govmmIch.Config
+	arch 	   ichArchBase
+}
+
+var ichKernelParams = []Param{
+	
+	{"root", "/dev/vda1"},
+	{"panic", "1"},
+	{"init", "/usr/lib/systemd/systemd"},
+	{"kvm-intel.nested", "1"},
+	{"no_timer_check" , ""},
+	{"noreplace-smp", ""},
+	{"page_alloc.shuffle", "1"},
+
+}
+
+var ichDebugKernelParams = []Param{
+	
+	{"console", "ttyS0,115200n8"},
+	{"systemd.log_level","debug"},
+	{"initcall_debug",""},
+	
+}
+
+//#####################################
+//
+// hypervisor interface implementation
+//
+//#####################################
+
+func (ich *cloudHypervisor) createSandbox(ctx context.Context, id string, networkNS NetworkNamespace, hypervisorConfig *HypervisorConfig, vcStore *store.VCStore) error {
+	ich.ctx = ctx
+
+	span, _ := ich.trace("createSandbox")
+	defer span.Finish()
+	
+	err := hypervisorConfig.valid()
+	if err != nil {
+		return err
+	}
+
+	ich.id = id
+	ich.store = vcStore
+	ich.config = *hypervisorConfig
+	ich.arch = newIchArch(ich.config)
+	ich.state.set(ichNotReady)
+	 
+	socketPath, err := ich.vsockSocketPath(id)
+	if err != nil {
+			ich.Logger().Info("Invalid socket path for cloud-hypervisor")
+			return nil
+	}
+	ich.socketPath = socketPath
+	
+	ich.Logger().WithField("function", "createSandbox").Info("creating Sandbox")
+
+	// TODO cross-check this with qemu procedures
+
+	// No need to return an error from there since there might be nothing
+	// to fetch if this is the first time the hypervisor is created.
+	if err := ich.store.Load(store.Hypervisor, &ich.info); err != nil {
+		ich.Logger().WithField("function", "createSandbox").WithError(err).Info("No info could be fetched")
+	}
+	
+	ichPath, err := ich.ichPath()
+	if err != nil {
+		return err
+	}
+	
+	// Add memory to the cloud hypervisor
+	memory, err := ich.arch.appendMemory(ich.config.MemorySize, "/dev/shm")
+	if err != nil {
+		return err
+	}
+
+	// Add vcpu to the cloud hypervisor
+	vcpu, err := ich.arch.appendProcessors(ich.config.NumVCPUs)
+	if err != nil {
+		return err
+	}
+	
+	// Add the kernel parameters and kernel path
+	kernelPath, err := ich.config.KernelAssetPath()
+	if err != nil {
+		return err
+	}
+	kernel := govmmIch.Kernel{
```

Comment:
Another comment here: Similar to the arch abstraction, I think the govmm implementation is also modeled after the current qemu implementation.
For Cloud Hypervisor, I don't think there's a need to go through a  govmm abstraction here. The `cloudHypervisor` structure would have a local array of `cloudHypervisorDevice`, etc.

The Cloud Hypervisor API is simple enough that we don't need to "hide" its complextiy behind the govmm package, and add a dependency between the Kata support for Cloud Hypervisor and the `govmm` package. This will simplify and speed the development process.